drivers/soundwire/intel.c

```diff
@@ -446,6 +446,44 @@ static int intel_link_power_down(struct sdw_intel *sdw)
 	return 0;
 }
 
+void sdw_intel_process_wakeen_event(struct sdw_intel_ctx *ctx)
+{
+	struct sdw_intel_link_res *link;
+	struct sdw_slave *slave;
+	struct sdw_bus *bus;
+	u16 wake_sts;
+
+	wake_sts = intel_readw(ctx->links->shim, SDW_SHIM_WAKESTS);
+
+	list_for_each_entry(link, &ctx->link_list, list) {
+		struct sdw_intel *sdw;
+
+		bus = &link->cdns->bus;
+		sdw = cdns_to_intel(link->cdns);
+
+		if (!(wake_sts & BIT(sdw->instance)))
+			continue;
+
+		/* disable WAKEEN interrupt ASAP to prevent interrupt flood */
+		intel_shim_wake(sdw, false);
+
+		/*
+		 * wake up master and slave so that slave can notify master
```

Comment:
we're only waking up the slave here right? So the comment "wake up master ..." is not correct as you dont wake master up explicitly here?

Reply:
@RanderWang this piece of code is indeed troubling.

I am not sure it's quite right to test for the slave status when handling the wakeen events, since the IP may be lost and the slave status is meaningless until the bus has actually restarted.

I think what you were trying to do here is to only resume actual devices that can generate a wake and ignore both the devices that cannot generate a wake and the ghost devices present in ACPI tables only, but the use of the status is very dangerous and racy.

Reply:
> we're only waking up the slave here right? So the comment "wake up master ..." is not correct as you dont wake master up explicitly here?

yes, not explicitly, just implicitly.  If slave is waken up, the parent master should be waken up.

Reply:
> @RanderWang this piece of code is indeed troubling.
> 
> I am not sure it's quite right to test for the slave status when handling the wakeen events, since the IP may be lost and the slave status is meaningless until the bus has actually restarted.

If the slave is unattached before clock stop, in my opinion, It can't wake up master with clock stop. 
> 
> I think what you were trying to do here is to only resume actual devices that can generate a wake and ignore both the devices that cannot generate a wake and the ghost devices present in ACPI tables only, but the use of the status is very dangerous and racy.

yes, you got my idea. I don't want to wake up any ghost device.

Reply:
@plbossart  I create a PR: https://github.com/plbossart/sound/pull/9  to remove these check. There is no regression issue.  Do you  prefer this defensive code or no check ? Thanks!

Reply:
@RanderWang your idea of not waking up ghost devices does not work with this code.

when a device is probe, we read the properties, and that's how we set wake-capable.
	/* device is probed so let's read the properties now */
	if (slave->ops && slave->ops->read_prop)
		slave->ops->read_prop(slave);

In other words, even ghost devices are seen as capable of generating wakes.

Reply:
@plbossart    Now I check slave->dev_num to skip ghost devices, please check https://github.com/plbossart/sound/pull/11. Thanks!